examples/advanced/nce_example.py

```diff
@@ -0,0 +1,237 @@
+#!/usr/bin/python3
+"""
+Counterfactual explanation of a user query.
+
+Based on application on the knapsack problem of Korikov, A., & Beck, J. C. Counterfactual Explanations via Inverse Constraint Programming.
+
+Usecase:
+ 1) Some optimal solution x* is provided to the user by a constraint optimization solver.
+ 2) User enters a query to change solution variables x_i to a new value.
+    This extra set of constraints is used to generate a new optimal solution x_d
+ 3) Using inverse optimization, we calculate the values of constraints the objective function which result in x_d being the optimal solution.
+    Extra constraint: only values corresponding to foil items are allowed to change.
+ 4) The user is presented with the result.
+
+
+Given:
+    A constraint optimization problem (c, f, X) with an optimal solution x*
+    A set of foil constraints assigning a value to variables in x*, resulting in x_d
+Find:
+    d such that min(f(d,X)) == x_d and ||c - d|| is minimal
+
+
+The algorithm consists of 2 alternating problems: the master problem (MP), and the sub problem (SP)
+
+Algorithm:
+    1 S = {}
+    2 Solve MP to obtain optimal d*
+    3 Solve SP with d* as c to obtain x_0
+    4 if objective(x_d) > objective(x_0):
+        add x_0 to S
```

Comment:
something more about the intuition....

you iterate optimal solutions to ||c-d|| until you find one for which min(f == x_d), something like that?

Reply:
maybe expl master prob first? not sure...

Reply:
Added some intuition